pkg/scheduler/backend/queue/backoff_queue.go

```diff
@@ -0,0 +1,231 @@
+*/
+
+package queue
+
+import (
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/backend/heap"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+	"k8s.io/kubernetes/pkg/scheduler/metrics"
+	"k8s.io/utils/clock"
+)
+
+// backoffQueuer is a wrapper for backoffQ related operations.
+type backoffQueuer interface {
+	shouldBeBackingoff(podInfo *framework.QueuedPodInfo) bool
+	getBackoffTime(podInfo *framework.QueuedPodInfo) time.Time
+	forEachBackoffCompleted(logger klog.Logger, fn func(pInfo *framework.QueuedPodInfo))
+
+	podInitialBackoffDuration() time.Duration
+	podMaxBackoffDuration() time.Duration
+
+	add(logger klog.Logger, pInfo *framework.QueuedPodInfo)
+	update(newPod *v1.Pod, oldPodInfo *framework.QueuedPodInfo) *framework.QueuedPodInfo
+	delete(pInfo *framework.QueuedPodInfo)
+	popBackoff() (*framework.QueuedPodInfo, error)
+	popErrorBackoff() (*framework.QueuedPodInfo, error)
+	get(pInfoLookup *framework.QueuedPodInfo) (*framework.QueuedPodInfo, bool)
+	has(pInfo *framework.QueuedPodInfo) bool
+	list() []*framework.QueuedPodInfo
+	len() int
+}
+
+// backoffQueue implements backoffQueuer and wraps two queues inside,
+// providing seamless access as if it were one queue.
+type backoffQueue struct {
+	clock clock.Clock
+
+	// podBackoffQ is a heap ordered by backoff expiry. Pods which have completed backoff
+	// are popped from this heap before the scheduler looks at activeQ
+	podBackoffQ *heap.Heap[*framework.QueuedPodInfo]
+	// podErrorBackoffQ is a heap ordered by error backoff expiry. Pods which have completed backoff
+	// are popped from this heap before the scheduler looks at activeQ
+	podErrorBackoffQ *heap.Heap[*framework.QueuedPodInfo]
+
+	podInitialBackoff time.Duration
+	podMaxBackoff     time.Duration
+}
+
+func newBackoffQueue(clock clock.Clock, podInitialBackoffDuration time.Duration, podMaxBackoffDuration time.Duration) *backoffQueue {
+	bq := &backoffQueue{
+		clock:             clock,
+		podInitialBackoff: podInitialBackoffDuration,
+		podMaxBackoff:     podMaxBackoffDuration,
+	}
+	bq.podBackoffQ = heap.NewWithRecorder(podInfoKeyFunc, bq.podsCompareBackoffCompleted, metrics.NewBackoffPodsRecorder())
+	bq.podErrorBackoffQ = heap.NewWithRecorder(podInfoKeyFunc, bq.podsCompareBackoffCompleted, metrics.NewBackoffPodsRecorder())
+
+	return bq
+}
+
+func (bq *backoffQueue) podInitialBackoffDuration() time.Duration {
+	return bq.podInitialBackoff
+}
+
+func (bq *backoffQueue) podMaxBackoffDuration() time.Duration {
+	return bq.podMaxBackoff
+}
+
+// podsCompareBackoffCompleted is a less function of podBackoffQ and podErrorBackoffQ.
+func (bq *backoffQueue) podsCompareBackoffCompleted(pInfo1, pInfo2 *framework.QueuedPodInfo) bool {
+	bo1 := bq.getBackoffTime(pInfo1)
+	bo2 := bq.getBackoffTime(pInfo2)
+	return bo1.Before(bo2)
+}
+
+// shouldBeBackingoff returns true if a pod is still waiting for its backoff timer.
+// If this returns true, the pod should not be re-tried.
```

Comment:
nit: can you have these kinds of comments within `type backoffQueuer interface {...}` too?